File: DetailViewTabCount/WinForms/VB/DetailViewTabCountVB.Module.Win/Controllers/EmployeeDetailViewWinController.vb
Imports System.ComponentModel
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Editors
Imports DevExpress.ExpressApp.Model
Imports DevExpress.ExpressApp.Win.Layout
Imports DevExpress.XtraLayout

Namespace DetailViewTabCountVB.Module.Win.Controllers
    Public Class EmployeeDetailViewWinController
        Inherits ObjectViewController(Of DetailView, Employee)
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            View.DelayedItemsInitialization = False
            AddHandler CType(View.LayoutManager, WinLayoutManager).ItemCreated, AddressOf DetailViewTabCountController_ItemCreated
        End Sub
        Private Sub DetailViewTabCountController_ItemCreated(ByVal sender As Object, ByVal e As ItemCreatedEventArgs)
            If TypeOf e.Item Is LayoutGroup AndAlso TypeOf e.ModelLayoutElement.Parent Is IModelTabbedGroup Then
                Dim layoutGroup = DirectCast(e.Item, LayoutGroup)
                For Each item In TryCast(e.ModelLayoutElement, IModelLayoutGroup)
                    Dim layoutViewItem As IModelLayoutViewItem = TryCast(item, IModelLayoutViewItem)
                    If layoutViewItem Is Nothing Then
                        Continue For
                    End If
                    Dim propertyEditor As ListPropertyEditor = TryCast(View.FindItem(layoutViewItem.ViewItem.Id), ListPropertyEditor)
                    If propertyEditor IsNot Nothing Then
                        propertyEditor.Frame.GetController(Of NestedListViewTabCountController)().Initialize(layoutGroup)
                    End If
                Next item
            End If
        End Sub
        Protected Overrides Sub OnDeactivated()
            RemoveHandler CType(View.LayoutManager, WinLayoutManager).ItemCreated, AddressOf DetailViewTabCountController_ItemCreated
            MyBase.OnDeactivated()
        End Sub
    End Class

    Public Class NestedListViewTabCountController
        Inherits ViewController(Of ListView)
        Private layoutGroup As LayoutGroup
        Public Sub New()
            TargetViewNesting = Nesting.Nested
        End Sub
        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            UpdateLayoutGroupText()
            AddHandler View.CollectionSource.CollectionChanging, AddressOf CollectionSource_CollectionChanging
            AddHandler View.CollectionSource.CollectionChanged, AddressOf CollectionSource_CollectionChanged
            AddHandler View.CollectionSource.CollectionReloaded, AddressOf CollectionSource_CollectionReloaded
            SubscribeToListChanged()
        End Sub
        Private Sub SubscribeToListChanged()
            Dim bindingList = GetBindingList(View.CollectionSource.Collection)
            If TypeOf bindingList Is IBindingList Then
                AddHandler bindingList.ListChanged, AddressOf CollectionSourceBindingList_ListChanged
            End If
        End Sub
        Public Sub Initialize(ByVal layoutGroup As LayoutGroup)
            Me.layoutGroup = layoutGroup
            UpdateLayoutGroupText()
        End Sub
        Private Sub CollectionSource_CollectionChanging(ByVal sender As Object, ByVal e As EventArgs)
            UnsubscribeFromListChanged()
        End Sub
        Private Sub CollectionSource_CollectionChanged(ByVal sender As Object, ByVal e As EventArgs)
            UpdateLayoutGroupText()
            SubscribeToListChanged()
        End Sub
        Private Sub CollectionSource_CollectionReloaded(ByVal sender As Object, ByVal e As EventArgs)
            UpdateLayoutGroupText()
        End Sub
        Private Sub CollectionSourceBindingList_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
            UpdateLayoutGroupText()
        End Sub
        Private Sub UpdateLayoutGroupText()
            If layoutGroup IsNot Nothing Then
                Dim count As Integer = View.CollectionSource.GetCount()
                layoutGroup.Text = DetailViewControllerHelper.ClearItemCountInTabCaption(layoutGroup.Text)
                If count > 0 Then
                    layoutGroup.Text = DetailViewControllerHelper.AddItemCountToTabCaption(layoutGroup.Text, count)
                End If
            End If
        End Sub
        Protected Overrides Sub OnDeactivated()
            RemoveHandler View.CollectionSource.CollectionChanging, AddressOf CollectionSource_CollectionChanging
            RemoveHandler View.CollectionSource.CollectionChanged, AddressOf CollectionSource_CollectionChanged
            RemoveHandler View.CollectionSource.CollectionReloaded, AddressOf CollectionSource_CollectionReloaded
            UnsubscribeFromListChanged()
            MyBase.OnDeactivated()
        End Sub
        Private Sub UnsubscribeFromListChanged()
            Dim bindingList = GetBindingList(View.CollectionSource.Collection)
            If TypeOf bindingList Is IBindingList Then
                RemoveHandler bindingList.ListChanged, AddressOf CollectionSourceBindingList_ListChanged
            End If
        End Sub
        Private Shared Function GetBindingList(ByVal collection As Object) As IBindingList
            If TypeOf collection Is IBindingList Then
                Return DirectCast(collection, IBindingList)
            ElseIf TypeOf collection Is IListSource Then
                Dim listSource = DirectCast(collection, IListSource)
                Return TryCast(listSource.GetList(), IBindingList)
            End If
            Return Nothing
        End Function
    End Class
End Namespace

File: DetailViewTabCount/WinForms/CS/DetailViewTabCount.Module.Win/Controllers/EmployeeDetailViewWinController.cs
using System;
using System.ComponentModel;
using DetailViewTabCount.Module.BusinessObjects;
using DetailViewTabCount.Module.Helpers;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Editors;
using DevExpress.ExpressApp.Model;
using DevExpress.ExpressApp.Win.Layout;
using DevExpress.XtraLayout;

namespace DetailViewTabCount.Module.Win.Controllers {
    public class EmployeeDetailViewWinController : ObjectViewController<DetailView, Employee> {
        protected override void OnActivated() {
            base.OnActivated();
            View.DelayedItemsInitialization = false;
            ((WinLayoutManager)View.LayoutManager).ItemCreated += DetailViewTabCountController_ItemCreated;
        }
        private void DetailViewTabCountController_ItemCreated(object sender, ItemCreatedEventArgs e) {
            if(e.Item is LayoutGroup layoutGroup && e.ModelLayoutElement.Parent is IModelTabbedGroup) {
                foreach(var item in (IModelLayoutGroup)e.ModelLayoutElement) {
                    if(item is IModelLayoutViewItem layoutViewItem && View.FindItem(layoutViewItem.ViewItem.Id) is ListPropertyEditor propertyEditor) {
                        propertyEditor.Frame.GetController<NestedListViewTabCountController>().Initialize(layoutGroup);
                    }
                }
            }
        }
        protected override void OnDeactivated() {
            ((WinLayoutManager)View.LayoutManager).ItemCreated -= DetailViewTabCountController_ItemCreated;
            base.OnDeactivated();
        }
    }

    public class NestedListViewTabCountController : ViewController<ListView> {
        private LayoutGroup layoutGroup;
        public NestedListViewTabCountController() {
            TargetViewNesting = Nesting.Nested;
        }
        protected override void OnActivated() {
            base.OnActivated();
            UpdateLayoutGroupText();
            View.CollectionSource.CollectionChanging += CollectionSource_CollectionChanging;
            View.CollectionSource.CollectionChanged += CollectionSource_CollectionChanged;
            View.CollectionSource.CollectionReloaded += CollectionSource_CollectionReloaded;
            SubscribeToListChanged();
        }
        private void SubscribeToListChanged() {
            if(GetBindingList(View.CollectionSource.Collection) is IBindingList bindingList) {
                bindingList.ListChanged += CollectionSourceBindingList_ListChanged;
            }
        }
        public void Initialize(LayoutGroup layoutGroup) {
            this.layoutGroup = layoutGroup;
            UpdateLayoutGroupText();
        }
        private void CollectionSource_CollectionChanging(object sender, EventArgs e) {
            UnsubscribeFromListChanged();
        }
        private void CollectionSource_CollectionChanged(object sender, EventArgs e) {
            UpdateLayoutGroupText();
            SubscribeToListChanged();
        }
        private void CollectionSource_CollectionReloaded(object sender, EventArgs e) {
            UpdateLayoutGroupText();
        }
        private void CollectionSourceBindingList_ListChanged(object sender, ListChangedEventArgs e) {
            UpdateLayoutGroupText();
        }
        private void UpdateLayoutGroupText() {
            if(layoutGroup != null) {
                int count = View.CollectionSource.GetCount();
                layoutGroup.Text = DetailViewControllerHelper.ClearItemCountInTabCaption(layoutGroup.Text);
                if(count > 0) {
                    layoutGroup.Text = DetailViewControllerHelper.AddItemCountToTabCaption(layoutGroup.Text, count);
                }
            }
        }
        protected override void OnDeactivated() {
            View.CollectionSource.CollectionChanging -= CollectionSource_CollectionChanging;
            View.CollectionSource.CollectionChanged -= CollectionSource_CollectionChanged;
            View.CollectionSource.CollectionReloaded -= CollectionSource_CollectionReloaded;
            UnsubscribeFromListChanged();
            base.OnDeactivated();
        }
        private void UnsubscribeFromListChanged() {
            if(GetBindingList(View.CollectionSource.Collection) is IBindingList bindingList) {
                bindingList.ListChanged -= CollectionSourceBindingList_ListChanged;
            }
        }
        private static IBindingList GetBindingList(object collection) {
            switch(collection) {
                case IBindingList bindingList:
                    return bindingList;
                case IListSource listSource:
                    return listSource.GetList() as IBindingList;
                default:
                    return null;
            }
        }
    }
}
